splitgpg2/__init__.py

```diff
@@ -799,25 +813,131 @@ async def inquire_CIPHERTEXT(self, untrusted_args):
+            ``untrusted_sexp``, so the length of ``untrusted_sexp`` is checked
+            to be 1 more than the length of ``names``.
+            """
+            if len(names) + 1 != len(untrusted_sexp):
+                raise Filtered
+            count = 0
+            for name in names:
+                count += 1
+                self.check_letter_sexp(name, untrusted_sexp[count], bytes)
+
+        def validate_ciphertext_sexp(*, untrusted_sexp):
+            """
+            Check that the ``untrusted_sexp`` is a valid offer of a ciphertext
+            for decryption.
+            """
+            untrusted_sexp = self.check_letter_sexp(b'enc-val', untrusted_sexp, list)
+            if len(untrusted_sexp) < 2:
+                raise ValueError('No MPIs found')
+            untrusted_alg = untrusted_sexp[0]
+            if untrusted_alg == b'ecdh':
+                check_mpi_list((b's', b'e'), untrusted_sexp=untrusted_sexp)
+            elif untrusted_alg == b'rsa':
+                check_mpi_list((b'a',), untrusted_sexp=untrusted_sexp)
+            elif untrusted_alg == b'elg':
+                check_mpi_list((b'a', b'b'), untrusted_sexp=untrusted_sexp)
+            else:
+                raise ValueError("Unknown encryption algorithm")
+
+        return self.inquire_command_D(validate_ciphertext_sexp, untrusted_args=untrusted_args)
+
+    def inquire_command_D_KEYGEN(self, *, untrusted_args: bytes):
+        def validate_bits_len(untrusted_sexp):
+            untrusted_bits = self.check_letter_sexp(b'nbits', untrusted_sexp, bytes)
+            if len(untrusted_bits) != 4 or untrusted_bits < b'1024' or untrusted_bits > b'4096':
+                raise ValueError('bad length')
+            for untrusted_v in untrusted_bits:
+                if untrusted_v < 48 or untrusted_v > 57:
+                    raise ValueError('bad length')
+        def validate_keygen_sexp(*, untrusted_sexp):
+            """
+            Check that the ``untrusted_sexp`` is a valid set of key generation
+            parameters.
+            """
+            untrusted_sexp = self.check_letter_sexp(b'genkey', untrusted_sexp, list)
+            if len(untrusted_sexp) < 2:
+                raise ValueError('No key parameters')
+            untrusted_alg = untrusted_sexp[0]
+            if untrusted_alg == b'ecc':
+                if len(untrusted_sexp) != 3:
+                    raise ValueError('invalid elliptic curve parameters')
+                untrusted_curve = self.check_letter_sexp(b'curve', untrusted_sexp[1], bytes)
+                untrusted_flags = untrusted_sexp[2]
+                if untrusted_curve == b'Ed25519':
+                    if untrusted_flags != [b'flags', b'eddsa', b'comp']:
+                        raise ValueError('invalid Ed25519 parameters')
+                elif untrusted_curve == b'Ed448':
+                    if untrusted_flags != [b'flags', b'comp'] and \
+                        untrusted_flags != [b'flags', b'eddsa', b'comp']:
```

Comment:
Does the flags order matter? Given the example of 4f773095076d783e5a34a7c096405b3fac3bf56d, I'd prefer to be less strict about the flags. Like, include allowed flags, but don't force them all being set and don't force the order. This would also simplify this whole `ecc` section (just allow client to choose from `comp`, `nocomp`, `eddsa`, `djb-tweak` flags, perhaps limiting last two to specific curves only _if_ they are indeed invalid for others).